core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -919,8 +925,13 @@ private[spark] object Utils extends Logging {
       // created the directories already, and that they are secured so that only the
       // user has access to them.
       randomizeInPlace(getYarnLocalDirs(conf).split(","))
-    } else if (conf.getenv("SPARK_EXECUTOR_DIRS") != null) {
-      conf.getenv("SPARK_EXECUTOR_DIRS").split(File.pathSeparator)
+    } else if (isRunningInK8sContainer(conf)) {
+      // Randomizing the shuffle location in case of K8s so that all disk get fair changes to
+      // get selected.
+      randomizeInPlace(conf.getenv("SPARK_LOCAL_DIRS").split(","))
```

Comment:
I am still not sure if this really safe to return the randomized directory whenever this method is invoked. This configuration (spark.local.dir) and environment variables are meant to be immutable but this is changing it here for every call during runtime.

Since this relates to the code path in which we should be relatively more strict, I would suggest to thoroughly investigate all the related paths TBH since we're breaking the assumption.

Reply:
QQ: I see both `SPARK_LOCAL_DIRS` and `SPARK_EXECUTOR_DIRS` in use within k8s resource manager - do we not want to do this for `SPARK_EXECUTOR_DIRS` @dongjoon-hyun ?

Reply:
> QQ: I see both `SPARK_LOCAL_DIRS` and `SPARK_EXECUTOR_DIRS` in use within k8s resource manager - do we not want to do this for `SPARK_EXECUTOR_DIRS` @dongjoon-hyun ?

We want it, @mridulm . That was my first comment here.
- https://github.com/apache/spark/pull/37203#pullrequestreview-1045944162